### PR TITLE
State dictionary retrieval from offloaded modules

### PR DESCRIPTION
**Pull request by blbadger**
# What does this PR do?

Adds a function to retrieve a state dictionary from an offloaded module, loading the parameters into a specified device.  

See transformers PR [#27412](https://github.com/huggingface/transformers/pull/27412) for more information

Working on the test too:)

TODO: replace disk offload conditional check with `expand_device_map`

@SunMarc

### CODE REVIEW

File: src/accelerate/utils/modeling.py [hunk 1]
```diff
@@ -1536,6 +1536,52 @@ def get_state_dict_offloaded_model(model: nn.Module):
...
+
+    Args:
+        module: (`torch.nn.Module`):
+            The module we want to retrieve a state dictionary from
+        module_name: (`str`):
+            The name of the module of interest
+        state_dict (`Dict[str: Union(str, torch.Tensor)]`):
+            Dictionary of {module names: parameters}
+        device_to_put_offload ('cpu' or 'gpu'):
+            Device to load offloaded parameters into, defaults to the execution device if not provided.
```
Comment (SunMarc) [inline src/accelerate/utils/modeling.py:1556]:
```suggestion
        device_to_put_offload (Union[int, str, torch.device], *optional*):
            Device to load offloaded parameters into, defaults to the execution device if not provided.
```

File: src/accelerate/utils/modeling.py [hunk 2]
```diff
@@ -1536,6 +1536,52 @@ def get_state_dict_offloaded_model(model: nn.Module):
...
+    for m_key in module.state_dict():
+        params = module.state_dict()[m_key]
+        if (root + f".{m_key}") in state_dict:
+            state_dict[root + f".{m_key}"] = params
+
+    if preforward:
+        # in case _hf_hook is not re-initialized
+        if device_to_put_offload:
+            module._hf_hook.execution_device = original_device
+        module._hf_hook.post_forward(module, torch.tensor([]))
```
Comment (SunMarc) [inline src/accelerate/utils/modeling.py:1580]: same

**Review by SunMarc — COMMENTED** [review]
Thanks for the work @blbadger ! Let's add a test and try it with the transformers PR ! After we manage to make it work on accelerate and transformers, we can merge this! I left a few comments

File: src/accelerate/utils/modeling.py [hunk 3]
```diff
@@ -1536,6 +1536,52 @@ def get_state_dict_offloaded_model(model: nn.Module):
...
+    """
+    from ..hooks import AlignDevicesHook
+
+    root = module_name[: module_name.rfind(".")]  # module name without .weight or .bias
+    preforward = False
+    if hasattr(module, "_hf_hook") and isinstance(module._hf_hook, AlignDevicesHook) and module._hf_hook.offload:
+        # re-specify the execution device where onloaded params are sent
+        if device_to_put_offload:
+            original_device = module._hf_hook.execution_device
+            module._hf_hook.execution_device = device_to_put_offload
```
[the remark this answers is not in the document]
Comment (blbadger) [reply src/accelerate/utils/modeling.py]: Good catch, I have modifed the `device_to_put_offload` kwarg to default to cpu

**Comment by blbadger** [comment]
Thanks for the review @SunMarc ! I have implemented your suggestions and wrote the test, and refactored the `transformers` code to use this method. 

 I have written the execution device to switch to the cpu by default, although if you think it would be a better idea to keep this as a gpu (if available) I would be happy to change. The `transformers` test is a little more straightforward if the cpu is default, although I know that fitting code for a test i not a great design consideration:)

**Comment by blbadger** [comment]
Just a heads up, I got around to integrating this module with transformers PR #[27412](https://github.com/huggingface/transformers/pull/27412) and testing looks good for large models there too.

**Review by SunMarc — APPROVED** [review]
Awesome @blbadger  ! I will merge this asap after we merge the PR that fixes the CI ! I've tested this PR with transformers and it works great !

**Pull request merged**